### DISCUSSION TITLE
[General] failed to validate the transaction. reason: Violated validation rules: Touched unallowed storage slots

### DISCUSSION THREAD

**Discussion by lutr0** [zkSync-Community-Hub/zksync-developers#236] [General] [answered]
### Environment

Mainnet

### Provide a brief description of the functionality you're trying to implement.

I have a working paymaster implementation that accepts any ERC20 token as payment.

### What is the specific issue or error you're encountering?

The paymaster is unable to interact with any ERC20 upgradable contract

### Can you share the error messages or logs you're receiving, if any?

`failed to validate the transaction. reason: Violated validation rules: Touched unallowed storage slots: address 5aea5775959fbc2557cc8789bc1bf90a239d9a91, key: b53127684a568b3173ae13b9f8a6016e243e63b6e8ee1178d6a717850b5d6103`

the transaction I am sending:
```json
{
    "rawTx": {
        "from": "0x79A4c883f7654ea81E17848c8aCbfeCe7d9234d2",
        "to": "0x5aea5775959fbc2557cc8789bc1bf90a239d9a91",
        "data": "0x095ea7b30000000000000000000000004bba932e9792a2b917d47830c93a9bc79320e4f7000000000000000000000000000000000000000000000000000012309ce54000",
        "value": "0",
        "customData": {
            "paymasterParams": {
                "paymaster": "0x9702beC6668A94619a9c3cfef0e220512FbEEfbd",
                "paymasterInput": "0x949431dc0000000000000000000000005aea5775959fbc2557cc8789bc1bf90a239d9a91000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000600000000000000000000000000000000000000000000000000000000000000041b33586b037338364b5646fd4bbcc5ab4cefb319d9a8ce3a64d042ff051413aea45474b3edba87fd5c3eaddd659313cad5318cf40f055c95b021a003be5a399051b00000000000000000000000000000000000000000000000000000000000000"
            },
            "gasPerPubdata": 50000
        },
        "maxPriorityFeePerGas": "1500000000",
        "maxFeePerGas": "2000000000",
        "gasLimit": 808524
    },
    "networkId": 324
}
```

### Have you made any recent changes to the contract before encountering this issue?

N/A

### Are there any external libraries or contracts that your contract interacts with?

N/A

### Can you provide the relevant portions of your contract code where the issue is occurring?

N/A

### Have you tried to isolate the problem, and if so, what were the results?

The problem seems to be limited to upgradable ERC20 using a TransparentUpgradeableProxy like WETH (`0x5aea5775959fbc2557cc8789bc1bf90a239d9a91`)

I wonder if there as issue with the implementation of the [allowed slots](https://era.zksync.io/docs/reference/concepts/account-abstraction.html#extending-the-set-of-slots-that-belong-to-a-user), not allowing the proxy contract.

### What steps have you already taken to try to resolve the issue?

N/A

**Comment by MexicanAce** [comment]
Hi @lutr0 👋 

Could you please link us to the code for the smart contract for us to investigate further?

**Reply by lutr0** [reply]
Sure thing, you can check this version of the repo: [https://github.com/ondefy/zyfi-paymaster/tree/cd61c97078e42baffb2972c5de72c4d7b07713e5](https://github.com/ondefy/zyfi-paymaster/tree/cd61c97078e42baffb2972c5de72c4d7b07713e5)

**Reply by MexicanAce** [reply]
Hey @lutr0 , apologies for the delayed response.

I believe the reason you're seeing this error is because the account logic can only touch slots that belong to the account. Please check https://era.zksync.io/docs/reference/concepts/account-abstraction.html#limitations-of-the-verification-step for further details on the limitations for verification.

**Reply by lutr0** [reply]
Hi @MexicanAce, thanks for checking this out!

I am still confused as I am performing a standard paymaster operation which works for all ERC20 apart from a few (e.g. WETH's TransparentUpgradeableProxy).
Does this mean that some ERC20 cannot be accepted a spayment by a paymaster? If so it may be worthwhile to adjust the verification logic.

**Reply by MexicanAce** [reply]
I can check with the protocol team to better understand the reasoning behind the verification logic (I know at a high-level it's for DOS protection), but I'll dig a little further into the TransparentUpgradeableProxy before doing so. Ideally, I want to get your code able to replicate the issue locally.

**Reply by StanislavBreadless** [reply, marked as answer]
EIP4337 (as well as our account abstraction protocol which is inspired by it) forbid accessing storage slots that do not belong to an account as it may lead to a situation when multiple transactions become invalid because of 1 storage slot change.

For instance, in the case of an upgradable proxy, if that proxy performs an upgrade, it may break all the accounts that rely on this contract in their validation logic. Our standard ERC20s that are bridged via the standard bridge are actually deployed as beacon proxies (and so in theory should fail the validation logic too), but they are whitelisted by the team (as they are ruled by the same governance as the protocol itself). But the AA protocol can not support any upgradable contract in general

**Reply by lutr0** [reply]
Thank you for the clarification, really appreciated.
It would be nice to have a clear documentation on this, so that upcoming projects creating tokens don't miss out on AA.

**Comment by loriyamazaki** [comment]
It seems that the issue you're encountering is related to the paymaster implementation's interaction with an upgradable ERC20 contract using a TransparentUpgradeableProxy. The error message indicates that the transaction failed to validate due to touching unallowed storage slots.

One possible explanation for this issue could be that the paymaster implementation is not configured to handle interactions with upgradable contracts or TransparentUpgradeableProxy contracts specifically. It's possible that the paymaster's validation rules do not allow interactions with certain storage slots, which may include those used by upgradable contracts.

To resolve this issue, you may need to update the paymaster implementation to properly handle interactions with upgradable contracts and TransparentUpgradeableProxy contracts. This may involve adjusting the validation rules to allow interactions with the relevant storage slots used by these types of contracts.

Additionally, you may want to review the implementation of the allowed storage slots in your paymaster contract to ensure that it includes the necessary slots for interacting with upgradable contracts and TransparentUpgradeableProxy contracts.

Overall, further investigation into the paymaster implementation's validation rules and interaction mechanisms with upgradable contracts will likely be necessary to resolve this issue.

**Comment by mganlok** [comment]
It seems like the issue might be related to the interaction between your paymaster implementation and upgradable ERC20 contracts deployed using TransparentUpgradeableProxy. The error message indicates that the paymaster is attempting to access unallowed storage slots, specifically referencing the proxy contract. Double-check your paymaster implementation to ensure it's compatible with proxies and handles storage slot access correctly. You may also need to review the allowed storage slots configuration in your contract. #Paymaster #ERC20 #ProxyContracts #StorageSlots

**Comment by sonhoranemichelle** [comment]
Ensure that the paymaster implementation allows interaction with upgradable ERC20 contracts through TransparentUpgradeableProxy.

**Comment by mapeshewa** [comment]
The paymaster implementation is unable to interact with any ERC20 upgradable contract and is encountering a "failed to validate the transaction" error with the reason being "Touched unallowed storage slots" for a specific address and key.
